In [1]:
import numpy as np

In [2]:
def f1(t, x, y, z):
    return y

def f2(t, x, y, z):
    return z

def f3(t, x, y, z):
    return x**2 - y**2 - (x - t)*z

In [52]:
t = 0.5
t_target = 2.2
h = 0.1
x = -1
y = 1
z = 2
X = [x]
Y = [y]
Z = [z]
X_hat = []
Y_hat = []
Z_hat = []

In [53]:
# The third-order adams-bashforth-moulton predictor-corrector method is not self-starting, 
# therefore we will use the forth-order Runge-Kutta method to take the inital step needed.

c_11 = h*f1(t, X[0], Y[0], Z[0])
c_21 = h*f2(t, X[0], Y[0], Z[0])
c_31 = h*f3(t, X[0], Y[0], Z[0])

c_12 = h*f1(t+h*1/2, X[0]+c_11*1/2, Y[0]+c_21*1/2, Z[0]+c_31*1/2)
c_22 = h*f2(t+h*1/2, X[0]+c_11*1/2, Y[0]+c_21*1/2, Z[0]+c_31*1/2)
c_32 = h*f3(t+h*1/2, X[0]+c_11*1/2, Y[0]+c_21*1/2, Z[0]+c_31*1/2)

c_13 = h*f1(t+h*1/2, X[0]+c_12*1/2, Y[0]+c_22*1/2, Z[0]+c_32*1/2)
c_23 = h*f2(t+h*1/2, X[0]+c_12*1/2, Y[0]+c_22*1/2, Z[0]+c_32*1/2)
c_33 = h*f3(t+h*1/2, X[0]+c_12*1/2, Y[0]+c_22*1/2, Z[0]+c_32*1/2)

c_14 = h*f1(t+h, X[0]+c_13, Y[0]+c_23, Z[0]+c_33)
c_24 = h*f2(t+h, X[0]+c_13, Y[0]+c_23, Z[0]+c_33)
c_34 = h*f3(t+h, X[0]+c_13, Y[0]+c_23, Z[0]+c_33)

X.append(X[0]+1/6*(c_11+2*c_12+2*c_13+c_14))
Y.append(Y[0]+1/6*(c_21+2*c_22+2*c_23+c_24))
Z.append(Z[0]+1/6*(c_31+2*c_32+2*c_33+c_34))

In [54]:
for i in range(1, int(np.round(t_target/h-t/h))):
    X_hat.append(X[i]+1/2*h*(3*f1(t+i*h, X[i], Y[i], Z[i])-f1(t+(i-1)*h, X[i-1], Y[i-1], Z[i-1])))
    Y_hat.append(Y[i]+1/2*h*(3*f2(t+i*h, X[i], Y[i], Z[i])-f2(t+(i-1)*h, X[i-1], Y[i-1], Z[i-1])))
    Z_hat.append(Z[i]+1/2*h*(3*f3(t+i*h, X[i], Y[i], Z[i])-f3(t+(i-1)*h, X[i-1], Y[i-1], Z[i-1])))

    X.append(X[i]+1/12*h*(5*f1(t+(i+1)*h, X_hat[i-1], Y_hat[i-1], Z_hat[i-1])+8*f1(t+i*h, X[i], Y[i], Z[i])-f1(t+(i-1)*h, X[i-1], Y[i-1], Z[i-1])))
    Y.append(Y[i]+1/12*h*(5*f2(t+(i+1)*h, X_hat[i-1], Y_hat[i-1], Z_hat[i-1])+8*f2(t+i*h, X[i], Y[i], Z[i])-f2(t+(i-1)*h, X[i-1], Y[i-1], Z[i-1])))    
    Z.append(Z[i]+1/12*h*(5*f3(t+(i+1)*h, X_hat[i-1], Y_hat[i-1], Z_hat[i-1])+8*f3(t+i*h, X[i], Y[i], Z[i])-f3(t+(i-1)*h, X[i-1], Y[i-1], Z[i-1])))    

In [56]:
print(f'x({t_target}) ~ X({t_target}) = {X[int(np.round(t_target/h-t/h))]}')
print(f'y({t_target}) ~ Y({t_target}) = {Y[int(np.round(t_target/h-t/h))]}')
print(f'z({t_target}) ~ Z({t_target}) = {Z[int(np.round(t_target/h-t/h))]}')

x(2.2) ~ X(2.2) = 2.923350357623869
y(2.2) ~ Y(2.2) = 1.9139910382868532
z(2.2) ~ Z(2.2) = -1.0484366320311291


Using a program derived from this algorithm with a step size $h = 0.1$ gives
$X(2.2) = 2.923350$ and, with a step size $h = 0.05$, gives $X(2.2) = 2.925418$. The
method of Richardson extrapolation given in Section 2.3.6, is equally applicable
to problems such as this one involving coupled equations. Since the third-order
predictor–corrector method used has a local error of $O(h^4)$, the global error will
be $O(h^3)$. The method therefore gives the estimated error in the second value of
$X(2.2)$ as $2.923350−2.925418)/7 = −0.000295$. For 6 decimal place accuracy in the
final estimate we need error $≤ 0.0000005$; in other words, the error must be reduced
by a factor of $0.000295/0.0000005 = 590$. Since we are using a third-order method,
the required step length will be $0.05/\sqrt[3]{
590} = 0.00596$. Rounding this down to a
suitable size suggests that a step size of $h = 0.005$ will give a solution accurate to
6 decimal places. In fact the program yields, with $h = 0.005$, $X(2.2) = 2.92575057$.
With $h = 0.0025$ it gives $X(2.2) = 2.92575089$. Richardson extrapolation predicts
the error in the $h = 0.0025$ solution as $0.000000046$ and therefore that in the $h = 0.005$ as $0.00000037$. The required accuracy was therefore comfortably
achieved using $h = 0.005$.